### PR TITLE
Autoscaling reactive storage decider

### PR DESCRIPTION
**Pull request by henningandersen**
This commits adds the initial version of the reactive storage decider.
The reactive storage decider will signal to scale up if storage alone
prevents allocation of unassigned shards or prevents shards from either
remaining where they are or moving to another node.

The reactive storage decider does not try to look into the future, thus
at the time the reactive decider asks to scale up, the cluster is
already in a need for more storage.

This initial commit will be followed up by:

- [ ] Integration test (once we can call the deciders through transport).
- [ ] ClusterInfo adjustments when the simulated start and allocate moves shards around.
- [ ] Better autoscaling context once we figure out how to
- [ ] Adapt to formalized tiers when available

### CODE REVIEW

File: x-pack/plugin/autoscaling/src/main/java/org/elasticsearch/xpack/autoscaling/Autoscaling.java [hunk 1]
```diff
@@ -141,7 +142,8 @@ boolean isSnapshot() {
         return List.of(
             new NamedWriteableRegistry.Entry(Metadata.Custom.class, AutoscalingMetadata.NAME, AutoscalingMetadata::new),
             new NamedWriteableRegistry.Entry(NamedDiff.class, AutoscalingMetadata.NAME, AutoscalingMetadata.AutoscalingMetadataDiff::new),
-            new NamedWriteableRegistry.Entry(AutoscalingDecider.class, AlwaysAutoscalingDecider.NAME, AlwaysAutoscalingDecider::new)
+            new NamedWriteableRegistry.Entry(AutoscalingDecider.class, AlwaysAutoscalingDecider.NAME, AlwaysAutoscalingDecider::new),
+            new NamedWriteableRegistry.Entry(ReactiveStorageDecider.class, ReactiveStorageDecider.NAME, ReactiveStorageDecider::new)
```
Comment (barkbay) [inline x-pack/plugin/autoscaling/src/main/java/org/elasticsearch/xpack/autoscaling/Autoscaling.java:146]:
```suggestion
            new NamedWriteableRegistry.Entry(AutoscalingDecider.class, ReactiveStorageDecider.NAME, ReactiveStorageDecider::new)
```

File: x-pack/plugin/autoscaling/src/main/java/org/elasticsearch/xpack/autoscaling/Autoscaling.java [hunk 2]
```diff
@@ -153,6 +155,11 @@ boolean isSnapshot() {
                 AutoscalingDecider.class,
                 new ParseField(AlwaysAutoscalingDecider.NAME),
                 AlwaysAutoscalingDecider::parse
+            ),
+            new NamedXContentRegistry.Entry(
+                ReactiveStorageDecider.class,
```
Comment (barkbay) [inline x-pack/plugin/autoscaling/src/main/java/org/elasticsearch/xpack/autoscaling/Autoscaling.java:160]:
```suggestion
                AutoscalingDecider.class,
```

**Review by barkbay — COMMENTED** [review]
I just did a quick test on ECK in order to familiarize myself with the new autoscaling APIs and to have a better understanding of the overall architecture, it works great ! 👍 _(I must confess that I was also happy to hack a little bit of Java after almost 2 years of golang)_

I got the expected response after filling up the storage:
```json
{
    "decisions": [
        {
            "reactive_storage": {
                "name": "reactive_storage",
                "type": "scale_up",
                "reason": "not enough storage available for unassigned shards"
            }
        }
    ]
}
```

I just had to amend this PR with 2 small modifications in order to make it work:
1. inject `ClusterInfoService`, `AllocationDeciders` and `ShardsAllocator` in `TransportGetAutoscalingDecisionAction`
2. update `TransportGetAutoscalingDecisionAction.masterOperation` to return the result.

I guess it is because this PR is still a wip as it is mentioned in the description.

**Comment by henningandersen** [comment]
Thanks @barkbay , will have a look at your comments. We intend to add the ability to get a decision out soon too (though in a separate PR).

**Comment by henningandersen** [comment]
Moved this work to a new PR in #65059 .

**Pull request closed without merging**